File: October 22/22 Oct/76. Minimum Window Substring LC.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        vector<int> t_count(60,0),window(60,0);
        
        for(char ch:t) t_count[ch-'A']++;
        
        int n=s.size(), matched=0, left=0;
        int windowSize=INT_MAX;
        char ch;
        string subString;
        for(int i=0;i<n;i++)
        {
            ch=s[i]-'A';
            window[ch]++;
            if(window[ch]<=t_count[ch])
                matched++;
            if(matched<t.size()) continue;
            ch=s[left]-'A';
            while(left<=i && window[ch]>t_count[ch])
            {
                left++;
                if(left>i)
                {
                    left--;
                    break;
                }
                window[ch]--;
                ch=s[left]-'A';
            }
            if((i-left+1)<windowSize)
            {
                windowSize=i-left+1;
                subString="";
                for(int j=left;j<=i;j++)
                    subString+=s[j];
            }
            
        }
        return subString;
    }
};

File: APRIL 2025/AMZN/76. Minimum Window Substring LC.java
class Solution {
    public String minWindow(String s, String t) {
        if(s.length()<t.length())
            return "";
        int[] t_freq = new int[60], window_freq = new int[60];

        for(char t_char:t.toCharArray())
            t_freq[t_char-'A']++;

        int N = s.length(), match = 0, index;
        int window_left=0, window_right=-1, window_size = N+5;

        for(int left=0, right=0; right<N; right++) {
            index = s.charAt(right)-'A';
            window_freq[index]++;
            
            if(window_freq[index]<=t_freq[index])
                match++;
            if(match < t.length())    
                continue; // this is a optimization, which starts moving left pointer only after capturing t string equivalent window

            index = s.charAt(left)-'A';
            while(left<=right && window_freq[index]>t_freq[index])
            {
                left++;
                if(left>right)
                {
                    left--;
                    break;
                }
                window_freq[index]--;
                index = s.charAt(left)-'A';
            }
            System.out.println(left+" "+right);
            if((right-left+1)<window_size)
            {
                window_size = (right-left+1);
                window_left = left;
                window_right = right; 
                System.out.println(window_left+" "+window_right);
            }
        }
        return s.substring(window_left, window_right+1);
    }
}
